### DISCUSSION TITLE
Chapters 2,3 - reorder and change titles

### DISCUSSION THREAD

**Discussion by rivantsov** [graphql/graphql-wg#1018] [General]
## Current Arrangement

![ChaptersTitles](https://user-images.githubusercontent.com/594969/173206836-5defe392-01c0-4f9e-8333-8f85a6033e46.jpg)

## Problem

1. The Language chapter -  11 out of 12 sections describe the Request language, that is about querying against the server's type system that is **NOT DESCRIBED YET** - it is the subject of chapter 2. You literally have to jump forward and back (pic) It seems more logical to describe type system first, then how to make requests against it
 
1. It might seem that 'Type System' would have no 'language' stuff, but - there is a lot of 'language' there! It defines the IDL (SDL). 

## Observation 
Indeed we have 2 languages in GraphQL: 
1. IDL/SDL
1. Request language 

They are similar, share some elements, but still different languages - their domains do not intersect (for a doc - it is one or another). 

## Suggestion 

New chapters order and titles: 

1. Type System and Schema Definition Language (SDL) 
1. Request Language 

Additional consideration: section '2.1 Source Text' in current version describes basic language elements and its lexic base (terminals). It is common to both languages (SDL and Request).
 In new arrangement, if we move the whole chapter it will end up in odd place - in the Request Language chapter (renamed from Language), after 'Type System and SDL'. It should retain the first position as it describes the common elements of languages. Maybe make it a separate chapter preceding these two? It is large enough to be a chapter.

**Comment by rivantsov** [comment]
@leebyron @benjie @IvanGoncharov @BoD @twof @michaelstaib @sachee @fotoetienne @yaacovCR @joeynenni
 and everybody else
Kindly inviting to discussion and requesting feedback on this important subject.
Thank you!

**Comment by nodkz** [comment]
I agree that the chapter may be reordered.
But in such a case chapters 3,4,5 should be moved together because they describe server-side stuff: 

<img width="193" alt="Screen Shot 2022-07-08 at 22 57 57" src="https://user-images.githubusercontent.com/1946920/178037444-af208880-50ea-486d-b5dd-d067a6b46d54.png">


---------

After my talks, I'm tired to explain to people that GraphQL has two languages – SDL (for schema definition and development) and query language (for runtime). 

Also, it will be nice to choose the exact terminology:
- SDL or IDL
- Query Language or Request Language

**Reply by rivantsov** [reply]
Thank you

about sections 3,4,5 moving together under 'type system' - not sure. 
yes, it is server stuff, but introspection is also about how to query it; so Request/Query language should be introduced at this point I think
Validation is about validation of requests, so request language should be described before that. 

About two languages - YES! thousand times yes! finally somebody agreeing with me on this! 

SDL vs IDL - IDL is a common name for this kind of languages and files. SDL (Schema Definition Languages) is GraphQL variant of IDL

Query vs Request language. - we recently started distinguish Query vs Request, carefully. Request -> Query or Mutation or Subscription, so clearly should be Request when we mean all three

**Reply by nodkz** [reply]
I think whole spec should be rearranged starting from splitting on three big units:
- type system
- server side staff
  - schema definition language
  - execution
  - validation
  - etc
- client side staff
  - request/operation language
  - introspection
  - query
  - mutations
  - etc

But I don't know is it worth it or not. I don't want disturb WG. Really for me stream and defer are important. But, really, I dont believe that they will land until the end of 2022. And discussing simultaneously a lot of different things overhelm WG.

**Reply by benjie** [reply]
> Query vs Request language. - we recently started distinguish Query vs Request, carefully. Request -> Query or Mutation or Subscription, so clearly should be Request when we mean all three

The term "operation" refers to a `query`, `mutation` or `subscription` operation; so if you want to say "a query, mutation or subscription operation" you can just say "an operation".

The term "request" (or "request for execution") encompasses a document (which may contain many operations) and a few other things (the schema and initial value (normally provided by the server internally), the variable values (if any) and the name of the operation to execute (optional unless there's more than one operation in the document)).

**Reply by rivantsov** [reply]
@nodkz 
I would disagree with this division: type-system, server, client. I think we should follow the rule that the story is linear, the later sections refer (require knowledge of) only earlier sections. With your order, for example, the Execution section in Server part, will involve knowledge of request language, but it is described in later chapter 'client side stuff'

**Reply by rivantsov** [reply]
@benjie 
I agree with what you say. But what we are looking for here is a simple 'label' to mark the sub-language; it does not have to be precise in terms of other specific definitions, but simply easy to remember and to associate with the sub-domain. 'Request language' is fine in this regard in my opinion; 'operation language' - fine too, although I would prefer the former (request).

**Reply by benjie** [reply]
The spec has the term ["ExecutableDocument"](https://spec.graphql.org/draft/#ExecutableDocument); perhaps that would be suitable?

**Reply by rivantsov** [reply]
@benjie 
you mean 'executable language'? Or maybe better 'execution language'; but I think still bit "misdirecting". Execution happens on the server side, inside the server, while the sub-language in point defines what's on-the-wire.

**Reply by benjie** [reply]
The ExecutableDocument is fed to GraphQL and GraphQL executes it... I don't think 'executable language' would be misdirecting. Please keep in mind that GraphQL-over-HTTP is not the only way GraphQL is used, e.g. in Gatsby it's used in the same process without any "on-the-wire" at all. Try not to think in terms of "client" and "server" because GraphQL itself isn't really defined in those terms; GraphQL is more of an execution engine and a query language for that execution engine independent of transport (if any).

**Reply by rivantsov** [reply]
Hi, @benjie sorry for late response, was away for a while
why did you see that I meant "over-HTTP"? 'on the wire' was just a short and simplified ref to 'communication channel' that transmits the GraphQL query text and response. I am well aware that there are cases when both client and server (service) components run in the same process on the same machine, and communication channel is just direct call from client to server. That's the case for Teams, and from speaking with my friends there - they still use terms client and server even when of course they know it's not over HTTP; these are just convenient shortcuts. 
> GraphQL is more of an execution engine and query language. 

Bang! query language! that's the term I would be happy with, much better than 'executable language'. But a bit misleading, since inside this language we distinguish queries, mutations, etc, would be a bit confusing. We started using 'request' term as a box for both, so Request Language seems still perfect name for it. 
As for 'GraphQL is execution engine' - still do not see it this way. Do we ship the engine? No. WG publishes spec; other people build conforming execution engines.

**Reply by benjie** [reply]
> As for 'GraphQL is execution engine' - still do not see it this way. Do we ship the engine? No. WG publishes spec; other people build conforming execution engines.

The GraphQL Specification is a specification that defines a "query language and execution engine". The "query language and execution engine" that the GraphQL Specification specifies is called "GraphQL". "GraphQL" is a query language and execution engine. The first sentence of the GraphQL Specification states as much:

> This is the specification for GraphQL, a query language and execution engine [...]

**Reply by rivantsov** [reply]
yeah, you are right, it states "This is the specification for GraphQL, a query language and execution engine [...]".

But it also states: 
> GraphQL is a query language designed to build client applications

so, what is it? The problem is that both of these are very problematic and referencing either of them for judgement on other issues is very problematic as well. The other discussion about formal GraphQL definition is aiming to fix that.

**Reply by twof** [reply]
I feel like saying it's confusing to just call GraphQL a query language without mentioning the executor is like saying it's confusing to call C++ a programming language without mentioning the compiler. Obviously there's a compiler. The details of the compiler don't actually matter to most people because most people are just trying to write some C++. If my audience was C++ compiler developers I might also mention the fact that C++ is processed by compilers, but I don't think it would be wrong to omit that fact. 

Perhaps considering the audience of the spec would be helpful. If we expect it to primarily be read by implementers, unconcerned with the use case of the things they're building, then maybe stripping out every mention of the historical and real world majority use case of GraphQL is a valid course of action. If that's not the expected audience of the spec, then maybe we don't do that. It could also be that different parts of the spec have different audiences, and the precision required in the Execution section isn't also required in the Overview section.

I'd also like to point out that the passage you're quoting could be speaking of history. GraphQL was literally designed to build client applications. That's not incorrect. It's now used for other things, but providing context for the purpose it was designed for is helpful.

Furthermore I feel that this whole discussion of "what is GraphQL exactly?" isn't particularly useful, and it's quite resource intensive. On the level of "Is a hotdog a sandwich". We're all [touching different parts of the elephant.](https://en.m.wikipedia.org/wiki/Blind_men_and_an_elephant)

**Reply by rivantsov** [reply]
Well, strangely this thread deviated from the original subject, which was - naming the chapters and finding a **convenient, clear term for part of GraphQL language that is used for formulating requests** (vs other part which is for describing schemas). The term, name of sublanguage (Request Language, Executable Language?) - does not need to be precise, or pin-pointing the philosophical question "What is GraphQL?" (we have a separate discussion for this). Just a convenient name, in context of 2 chapters, since each chapter defines its own sub-language.  

As for your reference to the 'Elephants' story, I do not think it works here at all. The blind men did not 'create' the elephant, or did not draft a spec for it. It was external thing, and of course personal impressions differ. We, the WG, on the other hand, we DEFINE the thing - GraphQL, and we must be precise and detailed about the thing we define (and control to some extent). Starting with basic definition "GraphQL is ...". 

We cannot say: well, what is GraphQL it's different thing for different people, opinions differ, and we are OK with anything, whatever. Everybody is entitled to his/her own beliefs. We cannot do that.

> GraphQL was literally designed to build client application.

Well, if it was a client application without server part? There were no Facebook GraphQL servers at the time, GraphQL was just client-side thing?
  
> Furthermore I feel that this whole discussion of "what is GraphQL exactly?" isn't particularly useful, and it's quite resource intensive.

I feel exactly opposite way. To create a spec for something without defining what it is - sounds really silly. And many of the discussions and disagreements here and around GraphQL come from this absence of clear definition and 'scope'.
